x/ratelimit/rate_limit.go

```diff
@@ -0,0 +1,155 @@
+package ratelimit
+
+import (
+	"encoding/json"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	transfertypes "github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+
+	ratelimitkeeper "github.com/Stride-Labs/stride/v4/x/ratelimit/keeper"
+	"github.com/Stride-Labs/stride/v4/x/ratelimit/types"
+)
+
+// Parse the denom from the Send Packet that will be used by the rate limit module
+// The denom that the rate limiter will use for a SEND packet depends on whether
+//    it was a NATIVE token (e.g. ustrd, stuatom, etc.) or NON-NATIVE token (e.g. ibc/...)...
+//
+// We can identify if the token is native or not by parsing the trace denom from the packet
+// If the token is NATIVE, it will not have a prefix (e.g. ustrd),
+//    and if it is NON-NATIVE, it will have a prefix (e.g. transfer/channel-2/uosmo)
+//
+// For NATIVE denoms, return as is (e.g. ustrd)
+// For NON-NATIVE denoms, take the ibc hash (e.g. hash "transfer/channel-2/usoms" into "ibc/...")
+func ParseDenomFromSendPacket(packet transfertypes.FungibleTokenPacketData) (denom string) {
+	// Determine the denom by looking at the denom trace path
+	denomTrace := transfertypes.ParseDenomTrace(packet.Denom)
+
+	// Native assets will have an empty trace path and can be returned as is
+	if denomTrace.Path == "" {
+		denom = packet.Denom
+	} else {
+		// Non-native assets should be hashed
+		denom = denomTrace.IBCDenom()
+	}
+
+	return denom
+}
+
+// Parse the denom from the Recv Packet that will be used by the rate limit module
```

Comment:
Given these are in the README, not sure it makes sense to repeat here. Maybe something like

A->B->C->A->C->B->A
A->B: A is acting as a source, B as a sink
B->C: B is a source, C is a sink
etc.

would be more helpful

Reply:
I'm not sure I see the argument for having less documentation in the code itself. It seems arguably more important to have this in the code than the README. Is your thinking just that the comment was overwhelming?

I clarified some of the description and added the A,B,C notation in the latest commit, let me know if that alleviates some of the concern